### PR TITLE
add operators.json with per-operator version, install method, and managed-namespace detection

### PR DESCRIPTION
**Pull request by pkoutsovasilis**
## Summary

Implements [#270](https://github.com/elastic/eck-diagnostics/issues/270).

Adds `operators.json` to the diagnostic archive - a JSON array with one entry per operator namespace, capturing the version, how the operator was installed, and which namespaces it manages. Also adds `namespaces.json` (full cluster namespace list) to the archive and emits a warning when the operator's managed namespaces are not fully covered by `--resources-namespaces`, so users know the diagnostic may be incomplete.

All detection is read-only: it inspects the operator StatefulSet or Deployment pod template spec (args, env, ConfigMap volumes) and resolves a single ConfigMap when a config file is mounted. No `kubectl exec` is required.

## Changed

- **`operators.json` added to archive** - a JSON array of operator objects, one per `--operator-namespaces` entry. Each object carries `namespace`, `version` (semver or `"unknown"`), `install_method` (`helm` / `yaml` / `olm` / `unknown"`), and `managed_namespaces` (`all: true` when unrestricted; `static: [...]` for a fixed list; `selector: {...}` for the enterprise label-selector).

- **`namespaces.json` added to archive** - full cluster namespace list, included at the root level alongside `nodes.json`.

- **Missing-namespace warning** - after resolving managed namespaces for every operator, the tool cross-references them against the namespaces being collected and logs a `WARNING` for any gap, directing the user to `--resources-namespaces`.

- **Install-method detection** - inferred from labels on the operator StatefulSet (`helm.sh/chart` → helm, `control-plane=elastic-operator` → yaml) or Deployment (`olm.owner` → olm).

- **Managed-namespace detection** - follows Viper's precedence (`--namespaces` arg > `NAMESPACES` env var > `namespaces:`/`namespace-selector:` in the config file). OLM's `ValueFrom.FieldRef` for `NAMESPACES` is resolved from the pod template annotations (where OLM injects `olm.targetNamespaces`). ConfigMap volume tracing handles directory mounts, `items` key→path remappings, and single-file `subPath` mounts. Warnings are emitted when the config source cannot be traced (e.g. mounted from a Secret) or when the ConfigMap cannot be read or parsed.

### CODE REVIEW

File: internal/operator.go [hunk 1]
```diff
@@ -0,0 +1,396 @@
...
+	var mountName, mountPath, relPath string
+	for _, vm := range container.VolumeMounts {
+		// SubPath mounts a single file: mountPath IS the file path, so require an exact match.
+		if vm.SubPath != "" && configPath == vm.MountPath {
+			mountName = vm.Name
+			mountPath = vm.MountPath
+			relPath = vm.SubPath
+			break
+		}
+		if vm.SubPath == "" && strings.HasPrefix(configPath, vm.MountPath) {
```
Comment (pebrc) [inline internal/operator.go]: `strings.HasPrefix(configPath, vm.MountPath)` has no path-boundary check. A mount at `/conf` would match a `configPath` of `/config/eck.yaml` (or `/conf-data/...`), because `/conf` is a literal string prefix of `/config`. When that happens the wrong mount is selected, `relPath` is left as the full path, and the resolved `dataKey` won't exist in the ConfigMap — it degrades to the "could not parse" warning + all-namespaces, so no crash, but detection is silently wrong.

Suggest requiring an exact match or a separator boundary, e.g. `configPath == vm.MountPath || strings.HasPrefix(configPath, strings.TrimSuffix(vm.MountPath, "/")+"/")`.

File: internal/operator.go [hunk 2]
```diff
@@ -0,0 +1,396 @@
...
+
+// warnMissingNamespaces logs a warning for each managed namespace not covered by collectedNamespaces.
+func warnMissingNamespaces(c *kubernetes.Clientset, infos []OperatorInfo, collectedNamespaces []string) {
+	collectedNamespacesSet := sets.New(collectedNamespaces...)
+
+	for _, info := range infos {
+		mn := info.ManagedNamespaces
+		switch {
+		case mn.All:
+			logger.Printf("WARNING: Operator in namespace %q manages all namespaces; the diagnostic only collects resources from %v. Specify additional workload namespaces via --resources-namespaces to broaden coverage.", info.Namespace, collectedNamespacesSet.UnsortedList())
```
Comment (pebrc) [inline internal/operator.go]: The default ECK install manages all namespaces, so this `WARNING` will fire on essentially every typical single-operator deployment, even when the user scoped the diagnostic intentionally. It's informative per the PR description, but likely to read as noise/alarm on the common case — consider softening the wording or lowering the level for the all-namespaces case.

Minor: `collectedNamespacesSet.UnsortedList()` yields a nondeterministic order in the message; sorting would make output stable.

File: internal/operator.go [hunk 3]
```diff
@@ -0,0 +1,396 @@
...
+	info := OperatorInfo{
+		Namespace:         namespace,
+		Version:           "unknown",
+		InstallMethod:     "unknown",
+		ManagedNamespaces: ManagedNamespaces{All: true},
+		parsedVersion:     fallbackMaxVersion,
+	}
+
+	if userSpecifiedVersion != "" {
+		if v, err := version.ParseSemantic(userSpecifiedVersion); err == nil {
```
Comment (pebrc) [inline internal/operator.go:53]: Minor regression: when a user-specified `--eck-version` fails `ParseSemantic`, it's now silently swallowed (Version stays `"unknown"`). The previous `detectECKVersion` logged the parse error here, which gave the user feedback that their input was malformed. Consider logging on the error branch.

File: internal/operator.go [hunk 4]
```diff
@@ -0,0 +1,396 @@
...
+type OperatorInfo struct {
+	Namespace         string            `json:"namespace"`
+	Version           string            `json:"version"`
+	ManagedNamespaces ManagedNamespaces `json:"managed_namespaces"`
+	InstallMethod     string            `json:"install_method"`
+	parsedVersion     *version.Version
+}
+
+// detectOperatorInfo gathers metadata about the ECK operator running in the given namespace.
+func detectOperatorInfo(c *kubernetes.Clientset, namespace, userSpecifiedVersion string) OperatorInfo {
```
Comment (pebrc) [inline internal/operator.go]: Test coverage: the extracted helpers (`detectInstallMethod`, `extractFlagValue`, `splitCSV`, `findConfigMapForPath`, `parseManagedNamespacesFromConfigMap`, `detectManagedNamespaces`, ...) are thoroughly table-tested — nice. The orchestration and version-extraction integration are not: `detectOperatorInfo`, `versionFromStatefulSet`, `podTemplateFromStatefulSet`/`podTemplateFromDeployment`, and `warnMissingNamespaces` have no tests. These need a fake clientset, which the repo already uses elsewhere. Worth at least one test that drives `detectOperatorInfo` end-to-end for the StatefulSet and OLM-Deployment paths (version + install method + managed namespaces together).

File: internal/operator.go [hunk 5]
```diff
@@ -0,0 +1,396 @@
...
+	annotationKey := annotationKeyFromFieldPath(src.FieldRef.FieldPath)
+	if annotationKey == "" {
+		return ""
+	}
+	return podMeta.Annotations[annotationKey]
+}
+
+// annotationKeyFromFieldPath extracts the annotation key from a fieldPath of the form
+// metadata.annotations['<key>'], returning "" for any other form.
+func annotationKeyFromFieldPath(fieldPath string) string {
```
Comment (pebrc) [inline internal/operator.go]: `annotationKeyFromFieldPath` reimplements a subset of the canonical Kubernetes downward-API fieldPath parser. `k8s.io/kubectl` (v0.36.2) is already a direct dependency, and it exports `pkg/util/fieldpath.SplitMaybeSubscriptedPath`, which does exactly this:

```go
// "metadata.annotations['olm.targetNamespaces']" -> ("metadata.annotations", "olm.targetNamespaces", true)
base, key, ok := fieldpath.SplitMaybeSubscriptedPath(fieldPath)
if !ok || base != "metadata.annotations" {
    return ""
}
return key
```

Reusing it deletes ~12 lines plus its test, handles edge cases the hand-rolled version doesn't (e.g. keys containing `[`/`]`), and would also cover `metadata.labels['...']` for free if ever needed. The hand-rolled version is correct for the only form OLM emits today (the annotation), so this is a maintainability/robustness suggestion rather than a bug — no new dependency required.

File: internal/operator.go [hunk 6]
```diff
@@ -0,0 +1,396 @@
...
+
+// extractFlagValue finds the value of --flag value or --flag=value in a list of args.
+func extractFlagValue(args []string, flag string) string {
+	for i, arg := range args {
+		if v, ok := strings.CutPrefix(arg, flag+"="); ok {
+			return v
+		}
+		if arg == flag && i+1 < len(args) {
+			return args[i+1]
+		}
```
Comment (pebrc) [inline internal/operator.go:206]: `extractFlagValue` returns the first match only, which diverges from how the operator actually parses `--namespaces`. That flag is a pflag `StringSlice`, so repeated occurrences are appended: `--namespaces=a --namespaces=b` yields `[a, b]` in the operator but only `[a]` here — a silent under-report of managed namespaces.

In practice Helm/OLM render the namespace list into the config file (or a single comma-separated flag), so repeated flags are uncommon and this is low severity. Worth either collecting all occurrences for the `--namespaces` case, or a short comment documenting that only the first occurrence is read.

(For the comma-separated single-flag form, `splitCSV` is fine — namespace names are DNS-1123 labels, so a plain comma split matches the operator's viper/mapstructure `StringToSliceHookFunc(",")` decoding exactly.)

**Review by pebrc — COMMENTED** [review]
Overall the direction looks good. The 🤖 found a few more things. I don't think any of them are blocking.

File: internal/operator.go (hunk 1 above)
Comment (pkoutsovasilis) [reply internal/operator.go]: valid point, fixed in da9cc855c095b5878aee60bd3dded71276751647

File: internal/operator.go (hunk 2 above)
Comment (pkoutsovasilis) [reply internal/operator.go]: indeed warning is an overstatement, changed in 25d8265732d2c6c4d40d0ca6687968dd977d96f3

File: internal/operator.go (hunk 3 above)
Comment (pkoutsovasilis) [reply internal/operator.go:53]: Added logging for the error in 3659dfaad8428e202d52075021a1c79b7ca1ce51

File: internal/operator.go (hunk 4 above)
Comment (pkoutsovasilis) [reply internal/operator.go]: ok sounds reasonable added unit-tests in f30bb2b8a30e7b40de1b2b859115814eacd4c11a

File: internal/operator.go (hunk 5 above)
Comment (pkoutsovasilis) [reply internal/operator.go]: TIL, changed in 4559ea3a6c9d291658d29f3d0e22108ec1de2af9

File: internal/operator.go (hunk 6 above)
Comment (pkoutsovasilis) [reply internal/operator.go:206]: nice catch, fixed in bf0928fce7f79bc1af770697845104dce46545e6

**Review by pebrc — APPROVED** [review]
LGTM
